### DISCUSSION TITLE
jooq:jooq-codegen-maven:3.19.16:generate failed: An API incompatibility was encountered while executing org.jooq:jooq-codegen-maven:3.19.16:generate

### DISCUSSION THREAD

**Discussion by markwimpory** [jOOQ/jOOQ#17780] [Q&A] [answered]
Im getting the following failure when trying to generate my jooq from an existing JPA entity class...

[ERROR] Failed to execute goal org.jooq:jooq-codegen-maven:3.19.16:generate (default-cli) on project csor-death-crindex: Execution default-cli of goal org.jooq:jooq-codegen-maven:3.19.16:generate failed: An API incompatibility was encountered while executing org.jooq:jooq-codegen-maven:3.19.16:generate: java.lang.NoSuchMethodError: 'void org.hibernate.tool.schema.spi.SchemaCreator.doCreation(org.hibernate.boot.Metadata, org.hibernate.tool.schema.spi.ExecutionOptions, org.hibernate.tool.schema.spi.SourceDescriptor, org.hibernate.tool.schema.spi.TargetDescriptor)'
[ERROR] -----------------------------------------------------
[ERROR] realm =    plugin>org.jooq:jooq-codegen-maven:3.19.16

It looks like that method on hibernate core is long gone.  Or am I missing something?

**Comment by lukaseder** [comment, marked as answer]
> It looks like that method on hibernate core is long gone. Or am I missing something?

Maybe, but our `jooq-meta-extensions-hibernate` module declares the right Hibernate dependency version. You shouldn't override it for the plugin.

**Reply by markwimpory** [reply]
Im not, Im assuming its using the version which is part of my main project.

**Reply by lukaseder** [reply]
Well you are, then. :) (`jooq-meta-extensions-hibernate` can't do much about the classpath of the project in which it is being used)

**Reply by markwimpory** [reply]
Im a bit confused, nothing in the docs mentions this, I just lifted from the docs the plugin config.  So are we concluding that it cannot be made to work as part of the same project.  If I know the version I could try and add that to the dependency of the plugin itself, but im not sure if that will work.  I guess as an alternative I could try and create a separate module for the plugin....maybe :)

**Reply by lukaseder** [reply]
> Im a bit confused, nothing in the docs mentions this

Well, as always with such things, I'd rather upgrade the Hibernate dependency, and thus fix this issue than document weird things. Just look at how many times they've broken compatibility in the past: https://stackoverflow.com/a/32227118/521799. I'd never be done documenting all the version specifics.

But since you've asked this question, there's now "documentation" in this issue :)

> So are we concluding that it cannot be made to work as part of the same project. If I know the version I could try and add that to the dependency of the plugin itself, but im not sure if that will work. 

Here's our current dependency version: https://github.com/jOOQ/jOOQ/blob/version-3.19.16/pom.xml#L58

> I guess as an alternative I could try and create a separate module for the plugin....maybe :)

Well, you could try adding an `<excludes>` section to the plugin dependencies to prevent your version from being loaded. But yes, creating a separate module would be the cleanest solution. Ideally:

- 1 module containing only entities
- 1 module for the code generation and generated code
- 1 module for your application

This is what jOOQ recommends: https://www.jooq.org/doc/latest/manual/code-generation/codegen-jpa/

See *"How to organise your dependencies"*.

**Reply by markwimpory** [reply]
I will give it a go, and also keep an eye out for a new version.
